### PR TITLE
Levent update

### PR DESCRIPTION
**Pull request by levoz92**
Removed redundant imports
Added optimization functions for data handling, tokenization, batch allocation and distribution of training.

### CODE REVIEW

**Review by musab-mk — CHANGES_REQUESTED** [review]
Could we remove the old code, as keeping them as comments is making it look not readable as much

**Comment by musab-mk** [comment]
Thank you for the PR btw

**Comment by levoz92** [comment]
@musab-mk Np. Check my new commit. I removed commented out code.

File: functionary/train/train.py [hunk 1]
```diff
@@ -248,6 +265,55 @@ def train():
         print_rank0("***** HERE ARE SOME EXAMPLES FROM EVALUATION ***")
         training_utils.print_some_examples(eval_dataset, tokenizer)
 
+    # Dynamic batch size based on max tokens per batch
```
Comment (musab-mk) [inline functionary/train/train.py:278]: While dynamic batch size might seem a good idea for dealing with memory issues, it would cause instabilities in training.  Due to the difference in gradient updates per conversation. I would rather prefer stability over memory efficiency. Stable updates with higher cost of GPUs should be preferred over a cheaper&faster training

File: functionary/train/train.py [hunk 2]
```diff
@@ -139,6 +134,23 @@ def trainer_save_model_safe(trainer: transformers.Trainer):
     trainer.save_model()
 
 
+"""
+Below is the updated train() function from LEVENT OZBEK.
```
Comment (musab-mk) [inline functionary/train/train.py:148]: Authorship tracking is a responsibility of git. We should remove all authorship info from the code.

File: functionary/train/train.py [hunk 3]
```diff
@@ -139,6 +134,23 @@ def trainer_save_model_safe(trainer: transformers.Trainer):
...
+- DataLoaders are constructed using BatchSampler to dynamically adjust the batch size per epoch.
+- distributed DataLoader is used if local_rank != -1.
+- updated to use the optimized preprocess_logits_for_metrics dynamically compute_metrics from training_utils.py.
+
+Advantages of These Changes:
+- handles datasets with varying sequence lengths dynamically
+- supports both single-GPU and distributed setups.
+"""
+
+
```
Comment (musab-mk) [inline functionary/train/train.py:163]: Updates would rather preferred to go into PR description, not the code

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: functionary/train/training_utils.py [hunk 4]
```diff
@@ -72,114 +121,78 @@ def initialize_tokenizer(
 
 
 def preprocess_logits_for_metrics(logits, labels, tokenizer_size):
-    """Preprocesses the logits during evaluation by computing the greedy token predictions for
-    accuracy calculation and loss values for perplexity calculation. Both pred_ids and loss are
-    of shape (batch_size x seq_len)"""
-
```
Comment (musab-mk) [inline functionary/train/training_utils.py:78]: We should keep docstring . If you think its obsolete in terms of information, better to update it, rather than removing completely.

File: functionary/train/training_utils.py [hunk 5]
```diff
@@ -72,114 +121,78 @@ def initialize_tokenizer(
...
+# Optimized compute metrics for large data sets!
+"""
+- Replaced loops over sequences with matrix operations.
+- NumPy for efficient batch-level calculations.
+- Moved evaluation computations to the GPU when possible.
+"""
+
+
 def compute_metrics(eval_preds, id2token, tokenizer):
-    """Computes next-token accuracy and perplexity metrics for evaluation"""
```
Comment (musab-mk) [inline functionary/train/training_utils.py:99]: Lets have a docstring here too

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: functionary/train/training_utils.py [hunk 6]
```diff
@@ -72,114 +121,78 @@ def initialize_tokenizer(
...
-    for token_id in dic:
+    # Token-specific accuracies
+    for token_id, stat in token_stats.items():
         token = id2token[token_id]
-        total_num = dic[token_id]["total"]
-        acc = -1
-        if total_num > 0:
-            acc = dic[token_id]["acc"] / total_num
-        metrics[f"accuracy_{token}"] = acc
-        metrics[f"accuracy_total_num_{token}"] = total_num
```
Comment (musab-mk) [inline functionary/train/training_utils.py:182]: I see some metrics has been removed. Could you kindly explain why these changes has been made?

**Review by musab-mk — CHANGES_REQUESTED** [review]
